src/raft_server.c

```diff
@@ -4915,131 +4915,188 @@ raft_server_net_client_request_init_sm_apply(
-                                                     reply_buf,
-                                                     reply_buf_sz);
-
-        rc_arr[i] = ri->ri_server_sm_request_cb(&rncr[i]);
+        // Move the offset to next entry
+        if(i > ri->ri_last_applied_sub_idx) {
+            struct raft_net_client_request_handle rncr;
+            raft_server_net_client_request_init_sm_apply(ri, &rncr,
+                                                        sink_buf + offset,
+                                                        reh.reh_entry_sz[i],
+                                                        reply_buf,
+                                                        reply_buf_sz);
+
+            rc = ri->ri_server_sm_request_cb(&rncr);
+            if (rc)
+                failed = true;
+        
+            // Increment the sub applied idx and store it in the
+            // persistent store.
+            raft_server_last_applied_increment(ri, &reh);
+            raft_server_sm_apply_opt(ri, &rncr.rncr_sm_write_supp);
+            raft_net_sm_write_supplement_destroy(&rncr.rncr_sm_write_supp);
+
+            /* Perform basic initialization on the reply buffer if the SM has
+            * provided the necessary info for completing the reply.  The SM
+            * would have called
+            * raft_net_client_request_handle_set_reply_info() if the necessary
+            * info was provided.  Note that the SM may not check for leader
+            * status, so the reply info may be provided even when this node
+            * is a follower.  Therefore, udp init should be bypassed if this
+            * node is not the leader.
+            */
+            if (!rc && raft_instance_is_leader(ri) &&
+                raft_net_client_request_handle_has_reply_info(&rncr) &&
+                uuid_compare(rncr.rncr_client_uuid,
+                            ri->ri_csn_this_peer->csn_uuid))
+            {
+                /* rncr and rcrm_data_size gets populated in
+                * ri_server_sm_request_cb. raft_server_client_reply_init()
+                * memset's rncr_reply which will overwrite the rcrm_data_size
+                * as well. So store the rcrm_data_size value in temporary variable
+                * first and restore it again.
+                */
+                struct raft_client_rpc_msg *reply = rncr.rncr_reply.rncr_reply_ptr;
+                uint32_t orig_rcrm_data_size = reply->rcrm_data_size;
 
-        if (rc_arr[i])
-            failed = true;
+                raft_server_client_reply_init(
+                    ri, &rncr, RAFT_CLIENT_RPC_MSG_TYPE_REPLY);
 
-        raft_net_sm_write_supplements_merge(&coalesced_ws,
-                                            &rncr_ptr->rncr_sm_write_supp);
+                reply->rcrm_data_size = orig_rcrm_data_size;
 
+                raft_server_reply_to_client(ri, &rncr, NULL);
+            }
+        }
+        
+        // Move the offset to next entry
```

Comment:
If this is placed into the `for` loop control then we can remove one layer of nesting in this loop by doing:
```
for (uint32_t i = 0; i < reh.reh_num_entries; offset += reh.reh_entry_sz[i], i++)
{
    if (i <= ri->ri_last_applied_sub_idx)
        continue;
        
    ...
}